### ISSUE TITLE
Ingame screenshot function

### ISSUE THREAD

**Issue by comnom** [endless-sky/endless-sky#3130]
I've added the ability to take a screenshot ingame here: https://github.com/endless-sky/endless-sky/compare/master...comnom:screenshot

I'm a bit torn as to whether this would be useful for enough people, so I decided to make an issue rather than a pull.

The implementation I have extends ImageBuffer with a FlipVertically() function and a SaveJPG() function. FlipVertically() is necessary because glReadPixels() reads from the frambuffer with 0,0 starting at the bottom left. SaveJPG() is our actual save function, and outputs the jpg at the default quality levels in libjpeg.

I also extended File/Files::Open() because we need to write the file in binary mode, which wasn't possible on Windows with the existing setup.

Screenshots get saved to config/screenshots, and the default key is F7. The key is arbitrary, and can be bound to something else, or unbound by default.

If this is something that gets added, ideally there should be a visual cue when you take a screenshot, but I'd rather handle that in a separate pull as I'm unfamiliar with that bit of code at the moment.

**Comment by Amazinite** [comment]
How would the functionality of this differ from Steam's screenshot system or a screenshot program such as Lightshot or Gyazo?

**Comment by comnom** [comment]
>How would the functionality of this differ from Steam's screenshot system or a screenshot program such as Lightshot or Gyazo?

It wouldn't, that's the rub. I no longer play through steam, and Windows built in screenshot capabilities are subpar. I'm not interested in third party software either, as I have no need for a screenshot program with a bunch of network strings attached, so I took this route.

**Comment by tehhowch** [comment]
Seems like you could drop some of the added functions in lieu of a default-false boolean for the existing Files::Open to get the binary write behavior, as everything seems pretty similar:
```
FILE *Files::Open(const string &path, bool write, bool writeBinary)
{
#if defined _WIN32
    return _wfopen(ToUTF16(path).c_str(), write ? (writeBinary ? L"wb" : L"w") : L"rb");
...
```

I think it'd be neat if you made the screenshots save as `<first> <last> YYYY-MM-DD hh:mm:ss.s` and then passed the in-game date and flagship as metadata to save in the file.
(I have minor dislike for writing image strings in such an order that they don't sort in time order when sorted by name :grinning: )

I'd like this function only for the automatic saving to a specific ES folder.

**Comment by comnom** [comment]
>Seems like you could drop some of the added functions in lieu of a default-false boolean...

Yea, I originally had it that way, but it seemed weird since writeBinary would only affect the Windows behavior. Really only makes sense if you know why Widows is a special snowflake and defaults to text mode. 

Easy enough to change if that's the preferred way of doing it. I did go further out of my way than I intended to get the mode specifier set up.

>I think it'd be neat if you made the screenshots save as `<first> <last> YYYY-MM-DD hh:mm:ss.s` and then passed the in-game date and flagship as metadata to save in the file.

I wouldn't mind doing the pilot name followed by yyyy-mm-dd-hh-mm-ss, so it's name/date sortable. Colon is reserved on Windows though.
